### DISCUSSION TITLE
STM32CubeMX Integration

### DISCUSSION THREAD

**Discussion by aprochazka** [hathach/tinyusb#1925] [Q&A]
Hello,
I am having trouble adding TinyUSB into my STM32 project generated by STM32CubeMX.
I am using Makefile to build and flash the project to my board (Nucleo-L432KC) and not IDE (STM32CubeIDE) and I can't get it to work following ideas inside common guides:
- [https://anvilelectronics.ovh/tinyusb-na-stm32-device-cdc/](https://web.archive.org/web/20211210013454/https://anvilelectronics.ovh/tinyusb-na-stm32-device-cdc/)
- https://github.com/hathach/tinyusb/discussions/633
- https://github.com/hathach/tinyusb/blob/master/docs/reference/getting_started.rst
- advices on various forums

My main trouble is to correctly edit Makefile to enable it to build, since Makefile generated by STM32CubeMX are not made for easy editing.

I have made following steps on clean project (https://github.com/aprochazka/TinyUSBCubeMX):
1. Create new project inside STM32CubeMX, enable USB communication with USB interrupts
2. Clone TinyUSB into subfolder in root of the project
3. Added tinyusb/src to include paths inside Makefile
4. Copied tusb_config.h from examples/device/cdc_msc/src to Core/Inc
5. Added to tusb_config.h as in https://github.com/hathach/tinyusb/discussions/633
```
#define CFG_TUSB_MCU                 OPT_MCU_STM32L4  
#define CFG_TUSB_OS                  OPT_OS_NONE  
#define BOARD_DEVICE_RHPORT_SPEED    OPT_MODE_HIGH_SPEED  // 480mbps   
#define BOARD_DEVICE_RHPORT_NUM     1  
#define CFG_TUSB_RHPORT1_MODE       (OPT_MODE_DEVICE | OPT_MODE_HIGH_SPEED) 
```

6. Copied usb_descriptors.c from examples/device/cdc_msrc/src to Core/Src
7. Included tusb.h in main.h `#include "tusb.h` and called `tusb_init()` in main.c
8. Inside stm32l4xx_it.c edited USB Interrupt handler to use `tud_int_handler()` function
```
void USB_IRQHandler(void)
{  
  tud_int_handler(1);
  return;
  HAL_PCD_IRQHandler(&hpcd_USB_FS);
}
```

After this project should to my knowledge still not work, but at least should build correctly, but make command returns this error message
```
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build/main.o: in function 'main':
/home/a/Projects/BP/TinyUSBTest/TinyUSBTest/Core/Src/main.c:94: undefined reference to 'tusb_init'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build/stm32l4xx_it.o: in function 'USB_IRQHandler':
/home/a/Projects/BP/TinyUSBTest/TinyUSBTest/Core/Src/stm32l4xx_it.c:207: undefined reference to 'dcd_int_handler'
```

I tried to simply copy the SRC_C from examples/rules.mk to C_SOURCES of makefile generated by CubeMX, but with no luck, since Makefile generated by CubeMX isn't built to be easily edited like this.

Is it even worth it to try and fix it or should I rather invest my time to port my project to another build toolchain, where TinyUSB is proven to work smoothly?

**Comment by hathach** [comment]
there is other topic on using cubeIDE with TinyUSB. It is not that hard if you understand how toolchain work e.g adding source file, include path, CFLAGS, and linker. Any IDE should allow you to specify these value in some sort of menu settings.
